enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,549 @@
+  cores for management services would offer exactly the same
+  performance as the default. It must be stable and continue to
+  operate reliably, but may respond more slowly.
+* This enhancement assumes that the configuration of management
+  services is done as part of installing the cluster, and cannot be
+  changed later. Future enhancements may address enabling or
+  reconfiguring the feature described here on an existing cluster.
+* This enhancement describes partitioning concepts that could be
+  expanded to be used for other purposes. Use cases for partitioning
+  workloads for other purposes may be addressed by future
+  enhancements.
+
+## Proposal
+
+[Previous attempts](https://github.com/openshift/enhancements/pull/62)
+to solve this problem focused on reducing or scaling requests so that
+the normal scheduling criteria could be used to safely place them on
+CPUs in the shared pool. This proposal reframes the problem, so that
+instead of considering "scaling" or "tuning" the requests we think about
+"isolating" or "partitioning" the workloads away from the non-management
+workloads. This view of the problem is more consistent with how the
+requirement was originally presented by the customer.
+
+We want to define "management workloads" in a flexible way. For the
+purposes of this document, "management workloads" include all
+OpenShift core components necessary to run the cluster, any add-on
+operators necessary to make up the "platform" as defined by telco
+customers, and operators or other components from third-party vendors
+that the customer deems as management rather than operational. It is
+important to note that not all of those components will be delivered
+as part of the OpenShift payload and some may be written by the
+customer or by vendors who are not our partners.
+
+The basic proposal is to provide a way to identify management
+workloads at runtime and to use CRI-O to run them on a user-selected
+set of CPUs, while other workloads will be prevented from running
+there. This effectively gives 3 pools of CPUs (shared, dedicated, and
+management) and means the shared CPU pool will need other CPUs to
+support burstable or best-effort workloads.
+
+We want the isolation to be in place from the initial buildout of the
+cluster, to ensure predictable behavior. Therefore, the feature must
+be enabled during installation. To enable the feature, the user will
+specify the set of CPUs to run all management workloads. When the
+management CPU set is not defined, the feature will be completely
+disabled.
+
+We generally want components to opt-in to being considered management
+workloads. Therefore, for a regular pod to be considered to contain a
+management workload it must be labeled with `io.openshift.management:
+true`.
+
+We want to treat all OpenShift components as management workloads,
+including those that run the control plane. Therefore, kubelet will be
+modified to treat all static pods as management workloads, when the
```

Comment:
Why?  Needs some justification vs just having the operators also write the label into static pods.

Reply:
kubelet will filter the pods with the label

Reply:
Was this a belt-and-braces solution, in case we missed a static pod? Or is this a vestigial idea from before we decided we would have to update all of the operators anyway?

Reply:
filter on kubelet side stops mirrors pods from inaccurately reporting resources back to kube-apiserver.

Reply:
Ah, right. We need to do the same mutation that we do for normal pods, to change the resource requests, because otherwise the static pods will appear to consume CPU resources on the host.

Reply:
my view is if you deploy a static pod to a node, its a management component in the sense that its not bound to scheduling constraint, starts independent of an apiserver, and requires root permission on the node to even lay down.

Reply:
i don’t agree.  i think there should be no magic, just the single label and the single path.  have label, go to cores.

Reply:
@smarterclayton there is no getting around carrying the filter patch in the kubelet admission of a static pod to strip/convert.  if your argument is 'only do that when the label is present' , i am fine with that.